### PR TITLE
feat: add context api

### LINKED ISSUE #61: Introduce Context
https://opentelemetry.io/docs/specs/otel/context/

### PR DESCRIPTION
**Pull request by hendriknielaender**
# Add OpenTelemetry Context API Implementation

## Overview

This PR introduces a robust, thread-safe, and spec-compliant implementation of the OpenTelemetry Context API.

https://opentelemetry.io/docs/specs/otel/context/

## Key Features

- **Immutable Contexts:**  
  Contexts are immutable by design, ensuring safe concurrent access and compliance with the OpenTelemetry specification. All mutation operations return a new context instance.

- **Thread-Local Context Stack:**  
  Implements a thread-local stack for implicit context propagation, following OpenTelemetry's LIFO attach/detach semantics. This enables context to flow transparently through the call stack without explicit parameter passing.

- **Extensive Test Coverage:**  
  Includes a suite of tests covering key creation, context immutability, stack operations, error handling, and chaining semantics.

## Usage Example

```zig
const context = @import("context");

// Compile-time key
const TraceIdKey = context.Key("trace_id");

// Attach a value to the current context
const allocator = ...;
const new_ctx = try context.getCurrentContext().setValue(allocator, TraceIdKey, .{ .string = "abc123" });
const token = try context.attachContext(new_ctx);
defer _ = context.detachContext(token) catch {};

// Retrieve the value later
const trace_id = context.getCurrentContext().getValue(TraceIdKey);
```

## API Summary

- `Context.init()` — Create an empty context
- `Context.getValue(key)` — Retrieve a value by key
- `Context.setValue(allocator, key, value)` — Add or update a value, returning a new context
- `Context.deinit()` — Release resources held by a context
- `Key(comptime name)` — Create a compile-time context key
- `createKey(name)` — Create a runtime context key
- `getCurrentContext()` — Get the current context for the thread
- `attachContext(context)` — Push a context onto the thread-local stack
- `detachContext(token)` — Pop a context from the stack using a token
- `cleanup()` — Clean up all thread-local context resources

## Next Steps

- Integrate with tracing APIs
- Expand test coverage for edge cases and concurrency

Closes #61

### CODE REVIEW

**Comment by inge4pres** [comment]
> @inge4pres maybe you can have a look here. I'm not sure with the further implementation in to the sdk. Is the current approach correct?

Thanks Hendrik!
I'll give it a pass, also cc'ing @kmos as the author of the issue.

**Comment by kmos** [comment]
@hendriknielaender I'll take a look, Thanks!

File: src/api/context/context.zig [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+/// Each call to `next()` returns a unique ID starting from 0.
+const ComptimeKeyGenerator = struct {
+    var next_id: usize = 0;
+
+    fn next() usize {
+        const id = next_id;
+        next_id += 1;
+        return id;
+    }
+};
```
Comment (inge4pres) [inline src/api/context/context.zig:34]: Not sure I get the exact purpose of this struct.
Are you stating that is thread-safe because it is owned by a `threadlocal` var (see next comment)?

Asking because there is no mutex or atomic so if `Key()`, which is marked as `pub` is consumed outside of `ContextStack`, there is a chance of a data race.

I wonder: what's the cost of having `next_id` as a `atomic.Value(usize)`, if it's only incremented uniquely in its own thread (ecause of the `threadlocal` at L#263.
Alternatively, marking `next` as `comptime` might also give us the same safety guarantee?

File: src/api/context/context.zig [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        }
+        var popped = self.contexts.pop().?;
+        (&popped).deinit();
+        return true;
+    }
+};
+
+/// Thread-local storage for the context stack.
+/// Each thread maintains its own independent context stack.
+threadlocal var context_stack: ?*ContextStack = null;
```
Comment (inge4pres) [inline src/api/context/context.zig:263]: Nailed it 👍🏼 
This is the same conclusion @kmos and I reached when reading the Context documentation for the first time last week.

File: src/api/context/context.zig [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    try std.testing.expect(try detachContext(token1));
+}
+
+test "compile time key creation" {
+    const MyKey = Key("my_service.request_id");
+    const OtherKey = Key("my_service.request_id");
+    try std.testing.expect(MyKey.id != OtherKey.id);
+    try std.testing.expectEqual(@as(usize, 0), MyKey.id);
+    try std.testing.expectEqual(@as(usize, 1), OtherKey.id);
+}
```
Comment (inge4pres) [inline src/api/context/context.zig:416]: Why is the test named "compile time"?
Don't we need to explicitly mark as `comptime` the assignments and function call to make them happen at comptime?

**Review by inge4pres — COMMENTED** [review]
Great work here 👏🏼 thanks for tackling this 🚀 

Can you add a test showing how the API works across threads boundaries?

File: src/api/context/context.zig (hunk 1 above)
Comment (kmos) [reply src/api/context/context.zig:34]: I totally agree with @inge4pres . my 2 cents I guess that the cost to have `next_id` as `atomic` should be reasonable.

Comment (hendriknielaender) [reply src/api/context/context.zig]: You are right. I added a comment; I hope that clarifies it.

**Comment by hendriknielaender** [comment]
> Great work here 👏🏼 thanks for tackling this 🚀
> 
> Can you add a test showing how the API works across threads boundaries?

Good point, added tests for this 🙏
